### ISSUE TITLE
Operation on text randomly merges words

### ISSUE THREAD

**Issue by Siguardius** [WordPress/gutenberg#11588]
**Describe the bug**
When interacting (bolding, inserting links) with text block that has pasted content with prior formatting, block will randomly merge a few words in the text.

**To Reproduce**
1. Copy & paste text into the text block directly from word editor (LibreOffice in my case)
2. Bold a word
3. Enjoy the bug

**Expected behavior**
A few worlds in the text block will be merged.

**Screenshots**
BEFORE:
![zrzut ekranu 2018-11-07 o 15 16 23](https://user-images.githubusercontent.com/11268816/48138609-f8b62280-e2a4-11e8-84e8-fa546a615ddc.png)
AFTER:
![zrzut ekranu 2018-11-07 o 15 16 31](https://user-images.githubusercontent.com/11268816/48138624-0075c700-e2a5-11e8-8e2e-f818c7bc0ab9.png)

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: Chrome
 - Version: 70.0.3538.77

**Additional context**
- Present in latest Gutenberg release
- Avoidable by pasting and matching the style

**Comment by iseulde** [comment]
Hi @Siguardius! I cannot reproduce the problem, we'll need some more information. Could you share what was logged in the browser console when you paste the text from LibreOffice?

**Comment by Siguardius** [comment]
@iseulde This is everything that was in the console.
[www.pixelophobia.pl-1541686251826.log](https://github.com/WordPress/gutenberg/files/2562073/www.pixelophobia.pl-1541686251826.log)

**Comment by designsimply** [comment]
Noting two additional reports of this from a duplicate at #11626 and their testing steps for reference:

> 1. Paste text containing soft line breaks (from Word or from a PDF)
> 1. Edit the pasted text
> 1. All instances of soft line breaks are now deleted

**Comment by swissspidy** [comment]
I just noticed the same when copying content which was on separate lines because each line was actually a `<div>`. So the HTML looks a bit like this:

```html
<div>Some Text <a href="https://example.com"><u>Some link</u></a></div><div>Some Text <a href="https://example.com"><u>Some link</u></a><br /></div><div>Some Text <a href="https://example.com"><u>Some link</u></a><br /></div><div>Some Text <a href="https://example.com"><u>Some link</u></a><br /></div><div>Some Text <a href="https://example.com"><u>Some link</u></a><br /></div>
```

The "Received plain text:" part in the console properly shows the content on individual lines.

The "Processed HTML piece:" section shows almost the identical HTML, but without the divs. Leading to the pasted content being on a single line.

**Comment by Haldaug** [comment]
Is it possible that the culprit of this bug could be these changes? https://github.com/WordPress/gutenberg/pull/10019

**Comment by Haldaug** [comment]
This bug can also be reproduced by manually inserting soft line breaks with "ctrl+shift" when you are editing paragraphs as HTML:

![gutenberg-line-break-bug](https://user-images.githubusercontent.com/3501914/48354475-2084fc00-e692-11e8-8e16-e69b77b536e4.gif)

**Comment by Haldaug** [comment]
I found a temporary fix for this. I added this code after line 218 in the [raw-handling](https://github.com/WordPress/gutenberg/blob/master/packages/blocks/src/api/raw-handling/index.js) code:

`piece = piece.replace( /\n/g, ' ' );`

It replaces all new lines with spaces. This is probably not ideal, but at least it lets you paste content from Word without randomly merging words again.

**Comment by iseulde** [comment]
Yes, we'll have to handle this in the raw handling and/or creating rich text values, maybe along with other non meaningful whitespace. I think what should happen is stripping leading and trailing non meaningful whitespace entirely, and collapsing multiple non meaningful white space characters into a single one.

**Comment by iseulde** [comment]
See also https://medium.com/@patrickbrosset/when-does-white-space-matter-in-html-b90e8a7cdd33.

What we have to do in rich text values seems to be: replace `[\n\t]` with spaces, then remove trailing and leading spaces, and reduce any other recurring spaces to one.

**Comment by designsimply** [comment]
Tested and confirmed using WordPress 4.9.8 and Gutenberg 4.3 by copying content from a word processing app to Gutenberg. I noticed the missing spaces match up with the line breaks in the Received HTML after pasting content:

![screen shot 2018-11-14 at 2 32 14 pm](https://user-images.githubusercontent.com/1119271/48513934-679df900-e81a-11e8-9915-4bc0ddf78272.png)

LibreOffice document I copied from:

<img src="https://user-images.githubusercontent.com/1119271/48512369-89957c80-e816-11e8-95e3-d090dc3be362.png" width="270">
<sup>Tested with LibreOffice 6.0.6.2 on macOS 10.13.6.</sup>

WordPress post I pasted into:

<img src="https://user-images.githubusercontent.com/1119271/48512541-f577e500-e816-11e8-8df5-15195cc54830.png" width="270">
<sup>Seen at http://alittletestblog.com/wp-admin/post.php?post=15818&action=edit running WordPress 4.9.8 and Gutenberg 4.3 using Firefox 63.0.1 on macOS 10.13.6.</sup>

Console output for reference:

```
Received HTML:

 <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
	<meta http-equiv="content-type" content="text/html; charset=utf-8"/>
	<title></title>
	<meta name="generator" content="LibreOffice 6.0.6.2 (MacOSX)"/>
	<style type="text/css">
		@page { margin: 0.79in }
		p { margin-bottom: 0.1in; direction: ltr; font-size: 11pt; line-height: 115%; text-align: left; orphans: 2; widows: 2 }
		a:link { color: #0000ff }
	</style>
</head>
<body lang="en-US" link="#0000ff" dir="ltr">
<p style="margin-bottom: 0in; font-weight: normal; line-height: 100%; page-break-after: avoid">
<font face="Arial, serif"><font size="3" style="font-size: 12pt"><span lang="en-CA">But
we </span></font></font><font face="Arial, serif"><font size="3" style="font-size: 12pt"><span lang="en-CA"><i>should</i></span></font></font><font face="Arial, serif"><font size="3" style="font-size: 12pt"><span lang="en-CA"><span style="font-style: normal">
know that mythology uses contradictory, or at least inconsistent,
versions of the same story, to express alternative perspectives in a
something else blah blah process, rather than “a lie”.</span></span></font></font></p>
<p style="margin-bottom: 0in; font-weight: normal; line-height: 100%">
<br/>

</p>
<p style="margin-bottom: 0in; font-weight: normal; line-height: 100%">
<font face="Arial, serif"><font size="3" style="font-size: 12pt"><span lang="en-CA"><span style="font-style: normal">The
importance of the death of Zeus is that the story emerges exactly from
that point in time and cultural transformation in which Zeus is also
born and at that time it was familiar for a vegetative god,
representing nature blooming in spring and dying in autumn, to die
and be re-born within the immortality of the eternal round of the
year or yearly daemon.</span></span></font></font></p>
</body>
</html>� 

Received plain text:

 But we should know that mythology uses contradictory, or at least inconsistent, versions of the same story, to express alternative perspectives in a something else blah blah process, rather than “a lie”.

The importance of the death of Zeus is that the story emerges exactly from that point in time and cultural transformation in which Zeus is also born and at that time it was familiar for a vegetative god, representing nature blooming in spring and dying in autumn, to die and be re-born within the immortality of the eternal round of the year or yearly daemon. index.js:50:160096

Processed HTML piece:
 <p>
But
we <em>should</em>
know that mythology uses contradictory, or at least inconsistent,
versions of the same story, to express alternative perspectives in a
something else blah blah process, rather than “a lie”.</p><p>
The
importance of the death of Zeus is that the story emerges exactly from
that point in time and cultural transformation in which Zeus is also
born and at that time it was familiar for a vegetative god,
representing nature blooming in spring and dying in autumn, to die
and be re-born within the immortality of the eternal round of the
year or yearly daemon.</p>
```

**Comment by iseulde** [comment]
I'll work on a fix asap.

**Comment by Haldaug** [comment]
FYI, my fix from here still works: https://github.com/WordPress/gutenberg/issues/11588#issuecomment-438270631

**Comment by iseulde** [comment]
Could anyone test #12093?

**Comment by designsimply** [comment]
Closed #12293 as a duplicate.

**Comment by designsimply** [comment]
I noticed that #12093 was superseded by #12166. I tested #12166 and left a comment with my findings (https://github.com/WordPress/gutenberg/pull/12166#issuecomment-441483884).

@sae13 the main way to test pull requests would be to install a local development environment as outlined in https://github.com/WordPress/gutenberg/blob/master/CONTRIBUTING.md#getting-started and then use `git` to checkout the branch for the pull request that needs testing. Setting up a local development environment wouldn't be expected unless you are contributing code or very deep into testing though. 🙂

I also found another way to test but you need to be logged in to GitHub and know the URL (_I think_). Go to https://github.com/WordPress/gutenberg/tree/try/rich-text-newlines-to-space and click the "Clone or download" button. That will give you a version of the plugin running the `try/rich-text-newlines-to-space` branch and you can delete Gutenberg and install+activate the downloaded one in order to test.

**Comment by gooma2** [comment]
Glad this is being worked on as it does this is every sentence pasted in. If you copy/paste to wordpad then it works fine, but having to do 2 steps with what was only 1 step before in Classic Mode seems counterproductive.

**Comment by iseulde** [comment]
The fix will be shipped in WordPress 5.0.1.

**Comment by sae13** [comment]
now button colors not working :rofl:

**Comment by MightyGadget** [comment]
> The fix will be shipped in WordPress 5.0.1.

Just to confirm that WordPress 5.0.1 has no fixed this. Copying from Word 2016 still has the problem

**Comment by gooma2** [comment]
Oddly enough, it actually has fixed the issue for me. I'm also able to
highlight blocks of words now unlike before.

The only difference now is that if I try to edit an older article, it won't
let me do it. It actually acts like the way 5.0 did when I tried to use
Gutenberg.

**Comment by swissspidy** [comment]
@mightygadget Since 5.0.1 is a security release only, it will be fixed 5.0.2 now. ETA is next week.

**Comment by MatildaMarseillaise** [comment]
I thought it was just me! I have upgraded to 5.0.1 and am finding that copying from Word into Wordpress draft article is also merging words. 

Also, if I am scheduling a post rather than posting it immediately it doesn't seem to be doing the automatic spell check (which I would hope would pick up these merged words). Any idea where the spellcheck function is now?

**Comment by gooma2** [comment]
I had that issue when updating to 5.0 but after 5.0.1 I was able to copy paste from Word into WP. Now it just won't let me edit older articles so back to Classic Editor. It'll get there.

**Comment by kylelarkin** [comment]
I can also confirm that this bug still exists in WordPress 5.0.1 with documents pasted from Word 2016. Copying text from a Word File into a Google doc and then into Gutenberg somehow bypasses the issue.

**Comment by jersnav** [comment]
It looks like some of my existing posts created before 5.0 are impacted by the missing space issue when the posts are converted to blocks. So the posts are correct until I go to edit the old posts and select "convert to blocks." At that point it seems like a few spaces in between words per paragraph disappear.

**Comment by ChapeauDigital** [comment]
I can also confirm that I get regular complains from my costumers, that pasting from Word is getting cropped of spaces between words.

**Comment by designsimply** [comment]
This issue should have been fixed in WordPress version 5.0.2 released on December 19, 2018. If you are still having trouble after updating to WordPress 5.0.2 or later, please leave a new comment there noting your WordPress version and any relevant details you think may be helpful (i.e. you are using a very old version of Word). Thank you!
